docs/proposals/004-initialization.md

```diff
@@ -0,0 +1,317 @@
+A type is a "legacy C-Style struct" if all of the following conditions are met:
+- It is a user-defined struct type or a basic scalar, vector or matrix type, e.g. `int`, `float4x4`.
+- It does not contain any explicit constructors defined by the user.
+- All its members have higher or equal visibility than the type.
+- All its members are legacy C-Style structs or arrays of legacy C-style structs.
+In such case, we perform a legacy "read data" style consumption of the initializer list, so that the following behavior is valid:
+
+```csharp
+struct Inner { int x; int y; };
+struct Outer { Inner i; Inner j; }
+
+Outer o = {1, 2, 3}; // Initializes `o` into `{ Inner{1,2}, Inner{3,0} }`.
+```
+
+If the type is not a legacy C-Style struct, Slang should produce an error.
+
+
+Examples
+-------------------
+```csharp
+
+// Assume everything below is public unless explicitly declared.
+
+struct Empty {}
+void test()
+{
+  Empty s0 = {}; // Works, `s` is considered initialized.
+  Empty s1; // `s1` is considered initialized bcause `Empty` is `IDefaultInitializable`.
+}
+
+struct CLike {int x; int y; }
+void test1()
+{
+  CLike c0; // `c0` is uninitialized.
+  CLike c1 = {}; // initialized with legacy initializaer list logic, `c1` is now `{0,0}`.
+  CLike c2 = {1}; // initialized with legacy initializaer list logic, `c1` is now `{1,0}`.
+  CLike c3 = {1, 2}; // initilaized with ctor call `CLike(1,2)`, `c3` is now `{1,2}`.
+}
+
+struct ExplicitCtor {
+   int x;
+   int y;
+   __init(int x) {...}
+}
+void test2()
+{
+  ExplicitCtor e0; // `e0` is uninitialized.
+  ExplicitCtor e1 = {1}; // calls `__init`.
+  ExplicitCtor e2 = {1, 2}; // error, no ctor matches initializer list.
+}
+
+struct DefaultMember {
+  int x = 0;
+  int y = 1;
+}
+void test3()
+{
+  DefaultMember m; // `m` is initialized to `{0, 1}`.
+  DefaultMember m1 = {1}; // calls `__init(1)`, initialized to `{1,1}`.
+  DefaultMember m2 = {1,2}; // calls `__init(1,2)`, initialized to `{1,2}`.
+}
+
+struct PartialInit {
+  // warning: not all members are initialized.
+  // members should either be all-uninitialized or all-initialized with
+  // default expr.
+  int x;
+  int y = 1;
+}
+void test4()
+{
+  PartialInit i; // `i` is not initialized.
+  PartialInit i1 = {2}; // calls `__init`, result is `{2,1}`.
+  PartialInit i2 = {2, 3}; // calls `__init`, result is {2, 3}
+}
+
+struct PartialInit2 {
+  int x = 1;
+  int y; // warning: not all members are initialized.
+}
+void test5()
+{
+  PartialInit2 j; // `j` is not initialized.
+  PartialInit2 j1 = {2}; // error, no ctor match.
+  PartialInit2 j2 = {2, 3}; // calls `__init`, result is {2, 3}
+}
+
+public struct Visibility1
+{
+  internal int x;
+  public int y = 0;
+}
+void test6()
+{
+  Visibility1 t = {0, 0}; // error, no matching ctor
+  Visibility1 t1 = {}; // error, no matching ctor
+}
+
+public struct Visibility2
+{
+  internal int x = 1;
+  public int y = 0;
+}
+void test7()
+{
+  Visibility2 t = {0, 0}; // error, no matching ctor
```

Comment:
From user-guide:

`internal` symbols are visible throughout the same module, regardless if it is referenced from the same type or same file. But they are not visible to other modules. 

so why is that?
Visibility2 t = {0, 0}; // error, no matching ctor 

are Visibility2 and test7 in the different module?

Reply:
This will fail because `Visiblity2` type contains members at different visiblity levels that disqualifies it from being a "C-style" struct type, and therefore we will NOT try the C-Style fallback logic to resolve the initializer list. If ctor match failed and the type is not c-style struct, it is an error.

Reply:
I see.